vnext/local-cli/runWindows/utils/autolink.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ * @format
+ */
+// @ts-check
+
+const execSync = require('child_process').execSync;
+const path = require('path');
+const fs = require('fs');
+
+function updateAutoLink(verbose) {
+  const execString =
+    'react-native config';
+  let output;
+  try {
+    console.log('Running react-natice config...');
+    output = execSync(execString).toString();
+    if (verbose)
+    {
+      console.log(output);
+    }
+
+    const config = JSON.parse(output);
+    const windowsPlatformConfig = config['project']['windows'];
+    const cppProjFile= windowsPlatformConfig?windowsPlatformConfig.cppProjFile:null;
+    if (cppProjFile == null){
+      console.log('AutoLink currently is only supported on cpp/winrt main project.');
+      return;
+    }
+    console.log('cppProjFile:' + cppProjFile);
+
+    const sourceDir = config['project']['windows'].sourceDir;
+    console.log('sourceDir:' + sourceDir);
+
+    const projectName = config['project']['windows'].projectName;
+    console.log('projectName:' + projectName);
+    
+    //#1. update nativeModules.g.h
+    const generatedHeader =  path.join(sourceDir, projectName, 'nativeModules.g.h');
+    if (!fs.existsSync(generatedHeader))
+    {
+      console.log('autoLink can not locate generated header file:' + generatedHeader);
+      return;
+    }
+
+    var generatedIncludes = '// NativeModules.g.h -- contents generated by "react-native run-windows"';
+    const dependencies = config['dependencies'];
+    var packageRegistrations = '#define RegisterNativeModulePackages';
+    for(const dependency in dependencies) {
+      const windowDependency = dependencies[dependency]['platforms']['windows'];
+      const cppProjFile = windowDependency?windowDependency.cppProjFile:null;
+      if (cppProjFile == null)
+      {
+        console.log('No cpp/winrt project found for ' + dependency);
+        continue;  
+      }
+      console.log('Adding include and package provider statement for ' + dependency);
+      const trimmedPackageName = windowDependency.packageName.trim();
+      const includeStatement = '\r\n#include ' + '<winrt/' + trimmedPackageName + '.h>';
+      generatedIncludes += includeStatement;
+      const packageRegistration = ' \\\r\n' + '    PackageProviders().Append(winrt::' + trimmedPackageName + '::ReactPackageProvider());';
```

Comment:
Consider using template strings instead of concatenation.

Reply:
Good to know!

---
In reply to: [391305475](https://github.com/microsoft/react-native-windows/pull/4293#discussion_r391305475) [](ancestors = 391305475)